Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift

```diff
@@ -0,0 +1,535 @@
+            }
+        }
+
+        mutating func http1Started() throws {
+            guard case .starting = self.state else {
+                throw Errors.connectionIsNotStarting
+            }
+
+            self.state = .http1(.idle(parked: false, idleSince: .now()))
+        }
+
+        mutating func park() throws {
+            switch self.state {
+            case .starting, .closed, .http1(.inUse), .http2(.inUse):
+                throw Errors.connectionNotIdle
+
+            case .http1(.idle(true, _)), .http2(.idle(_, true, _)):
+                throw Errors.connectionIsParked
+
+            case .http1(.idle(false, let lastIdle)):
+                self.state = .http1(.idle(parked: true, idleSince: lastIdle))
+
+            case .http2(.idle(let maxStreams, false, let lastIdle)):
+                self.state = .http2(.idle(maxConcurrentStreams: maxStreams, parked: true, lastIdle: lastIdle))
+            }
+        }
+
+        mutating func activate() throws {
+            switch self.state {
+            case .starting, .closed, .http1(.inUse), .http2(.inUse):
+                throw Errors.connectionNotIdle
+
+            case .http1(.idle(false, _)), .http2(.idle(_, false, _)):
+                throw Errors.connectionNotParked
+
+            case .http1(.idle(true, let lastIdle)):
+                self.state = .http1(.idle(parked: false, idleSince: lastIdle))
+
+            case .http2(.idle(let maxStreams, true, let lastIdle)):
+                self.state = .http2(.idle(maxConcurrentStreams: maxStreams, parked: false, lastIdle: lastIdle))
+            }
+        }
+
+        mutating func execute(_ request: HTTPSchedulableRequest) throws {
+            switch self.state {
+            case .starting, .http1(.inUse):
+                throw Errors.connectionNotIdle
+
+            case .http1(.idle(true, _)), .http2(.idle(_, true, _)):
+                throw Errors.connectionIsParked
+
+            case .http1(.idle(false, _)):
+                if let required = request.requiredEventLoop, required !== self.eventLoop {
+                    throw Errors.connectionDoesNotFulfillEventLoopRequirement
+                }
+                self.state = .http1(.inUse)
+
+            case .http2(.idle(let maxStreams, false, _)):
+                if let required = request.requiredEventLoop, required !== self.eventLoop {
+                    throw Errors.connectionDoesNotFulfillEventLoopRequirement
+                }
+                if maxStreams < 1 {
+                    throw Errors.connectionDoesNotHaveHTTP2StreamAvailable
+                }
+                self.state = .http2(.inUse(maxConcurrentStreams: maxStreams, used: 1))
+
+            case .http2(.inUse(let maxStreams, let used)):
+                if let required = request.requiredEventLoop, required !== self.eventLoop {
+                    throw Errors.connectionDoesNotFulfillEventLoopRequirement
+                }
+                if used + 1 > maxStreams {
```

Comment:
Seems odd to do the addition twice.
```suggestion
                if used >= maxStreams {
```